pkg/issuer/acme/setup_test.go

```diff
@@ -0,0 +1,520 @@
+
+		// Private key returned by keyFromSecret stub.
+		kfsKey crypto.Signer
+		// Error returned by keyFromSecret stub.
+		kfsErr error
+		// Whether keyFromSecret should be called.
+		kfsShouldBeCalled bool
+
+		// Whether RemoveClient should be called.
+		removeClientShouldBeCalled bool
+
+		// Whether AddClient should be called.
+		addClientShouldBeCalled bool
+
+		// Error returned by cl.Register
+		registerErr error
+
+		// ACME account returned by cl.GetReg
+		getRegAcc *acmeapi.Account
+		// Error returned by cl.GetReg
+		getRegErr error
+
+		acmePrivKeySecretCreateErr error
+
+		eabSecret       *corev1.Secret
+		eabSecretGetErr error
+
+		// expected ACME account passed to cl.Register
+		expectedRegisteredAcc *acmeapi.Account
+		// expected issuer conditions after Setup has been called.
+		expectedConditions []cmapi.IssuerCondition
+		expectedEvents     []string
+		wantsErr           bool
+	}{
+		"LetsEncrypt ACME v1 prod URL specified, return early": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEURL(acmev1Prod)),
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorInvalidConfig),
+					gen.SetIssuerConditionMessage(fmt.Sprintf(messageTemplateUpdateToV2, acmev1Prod, acmev2Prod))),
+			},
+		},
+		"LetsEncrypt ACME v1 staging URL with trailing slash specified, return early": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEURL(fmt.Sprintf("%s/", acmev1Staging))),
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorInvalidConfig),
+					gen.SetIssuerConditionMessage(fmt.Sprintf(messageTemplateUpdateToV2, fmt.Sprintf("%s/", acmev1Staging), acmev2Staging))),
+			},
+		},
+		"ACME private key secret does not exist, account key generation not disabled, key secret creation fails": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEPrivKeyRef(issuerSecretKeyName)),
+			kfsShouldBeCalled:          true,
+			kfsErr:                     notFoundErr,
+			acmePrivKeySecretCreateErr: someErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountRegistrationFailed),
+					gen.SetIssuerConditionMessage(messageAccountRegistrationFailed+someErr.Error())),
+			},
+			wantsErr: true,
+		},
+		"ACME private key secret does not exist, account key generation is disabled": {
+			issuer: gen.IssuerFrom(baseIssuer,
+				gen.SetIssuerACMEDisableAccountKeyGeneration(true),
+			),
+			kfsShouldBeCalled: true,
+			kfsErr:            notFoundErr,
+			expectedConditions: []cmapi.IssuerCondition{
+				*gen.IssuerConditionFrom(baseCondition,
+					gen.SetIssuerConditionReason(errorAccountVerificationFailed),
+					gen.SetIssuerConditionMessage(messageAccountVerificationFailed+messageNoSecretKeyGenerationDisabled+notFoundErr.Error())),
+			},
+			wantsErr: true,
+		},
```

Comment:
:+1:

Reply:
So the condition error message will be something like: `Failed to verify ACME account: the ACME issuer config has 'disableAccountKeyGeneration' set to true, but the secret was not found: Secret 'ns-1/secret-1' not found`.

Reply:
Is there a missing test for "ACME private key secret does not exist, account key generation is enabled, key secret creation succeeds" ?